tracks/python/exercises/matching-brackets/mentoring.md

```diff
@@ -0,0 +1,93 @@
+# Mentoring
+
+## Problem and Challenges
+
+The problem asks you verify that all brackets in a string are properly paired and nested.
+
+## Reasonable Solution
+
+```python
+def is_paired(input_string):
+    unmatched = []
+    matching = {")": "(", "]": "[", "}": "{"}
+    for character in input_string:
+        if character in matching.values():
+            # Push a new context
+            unmatched.append(character)
+        if character in matching:
+            if not unmatched:
+                return False
+            # Peek at the current context
+            if unmatched[-1] == matching[character]:
+                # Pop a context
+                unmatched.pop()
+            else:
+                return False
```

Comment:
Why is the peak necessary? 
```
if unmatched.pop() != matching[character]:
    return False
```
Short circuit evaluation could also be used here, though that might be pushing it for an 'easy' task?
```
if not unmatched or unmatched.pop() != matching[character]:
          return False
```

Reply:
Peaking isn't *necessary*. But it's a good solution.

Conceptually:
1. a pop occurs when a match is found. If there's no match, the value should not be popped. This allows, say, the remaining stack to be accurately printed or whatnot.
2. this exercise is all about using a stack and peak is part of the standard stack operations.

Reply:
Okay, makes sense!